bin/cylc-check-software

```diff
@@ -16,84 +16,239 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-# Check for cylc software dependencies
-
-usage() {
-    cat <<eof
-Usage: cylc [admin] check-software
-
-Check that the external software required by cylc is installed.
-
-Minimum versions are not checked, except in the case of Python.
-
-Options:
-  -h, --help   Print this help message and exit.
-eof
+"""
+    cylc-check-software
+    Check for external software and version dependencies of cylc.
+"""
+
+import sys
+import re
+from subprocess import check_call, PIPE, Popen, CalledProcessError
+
+
+# Standardised output messages
+FOUND_NOVER_MSG = 'FOUND'
+MINVER_MET_MSG = 'FOUND & min. version MET'
+MINVER_NOTMET_MSG = 'FOUND but min. version NOT MET'
+FOUND_UNKNOWNVER_MSG = 'FOUND but could not determine version (?)'
+NOTFOUND_MSG = 'NOT FOUND (-)'
+
+"""Specification of cylc core & full-functionality module requirements, the
+   latter grouped as Python, TeX or 'other' (neither). 'opt_spec' item format:
+   <MODULE>: [<MIN VER OR 'None'>, <FUNC TAG>, <GROUP>, <'OTHER' TUPLE>] with
+   <'OTHER' TUPLE> = (<BASE CMD(S)>, <VER OPT>, <REGEX>, <OUTFILE ARG>)."""
+req_py_ver_range = (2, 6), (3,)
+opt_spec = {
+    'requests': [(2, 4, 2), 'HTTPSCOMMS', 'PY'],
+    'OpenSSL': [None, 'HTTPSCOMMS', 'PY'],
+    'urllib3': [None, 'HTTPSCOMMS', 'PY'],
+    'pygtk': [(2, 0), 'GUIORGRAPH', 'PY'],
+    'pygraphviz': [None, 'GUIORGRAPH', 'PY'],
+    'texlive': [None, 'LATEXGUIDE', 'TEX'],
+    'tocloft': [None, 'LATEXGUIDE', 'TEX'],
+    'framed': [None, 'LATEXGUIDE', 'TEX'],
+    'preprint': [None, 'LATEXGUIDE', 'TEX'],
+    'tex4ht': [None, 'LATEXGUIDE', 'TEX'],
+    'TeX': [(3, 0), 'LATEXGUIDE', 'OTHER', ('tex', '-v', r'TeX ([^\s]+)')],
+    'graphviz': [None, 'GUIORGRAPH', 'OTHER',
+                 ('dot', '-V', r'graphviz version ([^\s]+)', 2)],
+    'imagemagick': [None, 'HTMLUGUIDE', 'OTHER',
+                    ('magick identify', '-version', r'ImageMagick ([^\s]+)')]
```

Comment:
Just spotted this one. The `magick` top level command was introduced in a recent version of ImageMagick. To discover older versions of ImageMagick, you should probably use `convert` or `identify` without the `magick ...` prefix.

Reply:
(good spotting)

Reply:
I agree, very well observed - in fact it turns out from checking those commands that I do actually have 'ImageMagick' installed locally, when I thought I did not when I was first writing this PR because I got a `command not found` with 'magick' prepended! I will make changes to allow both command variants, and checking of multiple command variants in general as future module dependencies may well need this utility too.